includes/utils/class-amp-validation-utils.php

```diff
@@ -231,14 +301,133 @@ public static function validate_content( $post ) {
 		}
 	}
 
+	/**
+	 * Wraps callbacks in comments, to indicate to the sanitizer which plugin added them.
+	 *
+	 * Iterates through all of the registered callbacks.
+	 * If a callback is from a plugin and outputs markup,
+	 * this wraps the markup in comments.
+	 * Later, the sanitizer can identify which plugin any illegal markup is from.
+	 *
+	 * @global array $wp_filter
+	 * @return void
+	 */
+	public static function callback_wrappers() {
+		global $wp_filter;
+		if ( ! self::do_validate_front_end() ) {
+			return;
+		}
+		$pending_wrap_callbacks = array();
+		foreach ( $wp_filter as $filter_tag => $wp_hook ) {
+			foreach ( $wp_hook->callbacks as $priority => $callbacks ) {
+				foreach ( $callbacks as $callback ) {
+					$plugin = self::get_plugin( $callback['function'] );
+					if ( isset( $plugin ) ) {
+						$pending_wrap_callbacks[ $filter_tag ][] = array_merge(
+							$callback,
+							compact( 'plugin', 'priority' )
+						);
+					}
+				}
+			}
+		}
+
+		// Iterate over hooks to replace after iterating over all to begin with to prevent infinite loop in PHP<=5.4.
+		foreach ( $pending_wrap_callbacks as $hook => $callbacks ) {
+			foreach ( $callbacks as $callback ) {
+				remove_action( $hook, $callback['function'], $callback['priority'] );
+				$wrapped_callback = self::wrapped_callback( $callback );
+				add_action( $hook, $wrapped_callback, $callback['priority'], $callback['accepted_args'] );
+			}
+		}
+	}
+
+	/**
+	 * Gets the plugin of the callback, if one exists.
+	 *
+	 * @param string|array $callback The callback for which to get the plugin.
+	 * @return string|null $plugin   The plugin to which the callback belongs, or null.
+	 */
+	public static function get_plugin( $callback ) {
```

Comment:
This function seems like it should be merged with `get_source`, and it can then return an array including the source type (e.g. `plugin`) and the source name (e.g. `jetpack`).

Reply:
Thanks, [this commit](https://github.com/Automattic/amp-wp/commit/66ab20d3cae169cec044ea26331bec2362eb0322) moves `get_plugin()` into `get_source()`. As you mentioned, it returns an array with the `'type'` and `'source'`.